[vera] ip

verakohh opened the pull request:
# Cleo: Your Personal Task Management Assistant

### Cleo is a command-line based task management application that helps users keep track of their tasks efficiently. Designed to be simple yet powerful, Cleo enables users to manage their to-dos, deadlines, and events through an intuitive and straightforward text interface.

## Features:
**Task Management**: Easily add, mark, unmark, and delete tasks.
**Categorisation**: Organise tasks as To-Dos, Deadlines, and Events, each with specific attributes like due dates and times.
**Persistent Storage**: Cleo saves all tasks to a file, ensuring that your data is retained between sessions.
**User Interaction**: Cleo provides helpful feedback with each command, making it easy for users to stay on top of their tasks.

## How It Works:
1. Add a Task:
   - Use commands like todo, deadline, and event to add new tasks to the list.
2. Manage Tasks:
   - Mark tasks as done, unmark them if needed, or delete them entirely.
3. View Tasks:
   - List all tasks, or find specific tasks by searching with keywords.

## Example commands
```
todo Read a book
deadline Submit assignment /by 2024-09-10
event Meeting /from 2024-09-11 10:00 /to 2024-09-11 11:00
list
mark 1
unmark 2
delete 3
find book
```
## Project Setup:
To run Cleo, clone the repository and follow these steps:

```sh
./gradlew build
./gradlew run
```

## Development:
- **Languages**: Java
- **Build Tool**: Gradle

## Task List:
- [x] Implement basic task management (ToDos, Deadlines, Events)
- [x] Add persistent storage support
- [x] Improve user interface and error handling

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

verakohh committed:
Rename main class and add greeting

verakohh committed:
Rename main class and add greeting

verakohh committed:
Add support for echo

verakohh committed:
added list

verakohh committed:
mark as done, unmark as done, and no tasks yet

verakohh committed:
Deadlines, Events, and ToDos

verakohh committed:
removed whitespaces and included automated text UI testing

verakohh committed:
added Exceptions

verakohh committed:
Added delete function with updated automatic testing

verakohh committed:
use enums

verakohh committed:
save the tasks in hard disk

verakohh committed:
Merge branch 'branch-Level-7'

verakohh committed:
Add dates and times

verakohh committed:
Merge branch 'branch-Level-8'

verakohh committed:
Make the code more OOP

verakohh committed:
Merge tag 'A-MoreOOP'

verakohh committed:
Divide classes into packages

verakohh committed:
Merge tag 'A-Packages'

verakohh committed:
Merge branch 'add-gradle-support'

verakohh committed:
Change build.gradle names

verakohh committed:
Add JUnit tests

verakohh committed:
Merge tag 'A-JUnit'

verakohh committed:
Package the App as a JAR file

verakohh committed:
Merge branch 'A-Jar'

verakohh committed:
Add JavaDoc comments

verakohh committed:
Merge branch 'branch-A-JavaDoc'

verakohh committed:
Tweak the code to comply with a coding standard

verakohh committed:
Give users a way to find a task by searching for a keyword

verakohh committed:
Merge commit '40c1941dfeb7361764bbf821e2c4e6345149ed6a'

verakohh committed:
Merge branch 'branch-A-CodingStandard'

verakohh committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/cleo/Cleo.java

verakohh committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/cleo/Cleo.java

verakohh committed:
Merge branch 'master' of github.com:verakohh/ip

verakohh committed:
Fix bug form merge

verakohh committed:
Update javadocs to proper coding standard

verakohh committed:
Update javadocs to coding standard.

davidgohzk reviewed (commented):
Overall, the naming is ok, but what is a little weird is in the Cleo.java file, in the switch case, there is a hybrid of where the System.out.println is found. For instance, in the case `UNMARK`, it is found in the switch case whereas in `DELETE`, it is found in the `tasks.removeTask`. This is not good as there is a lack of consistency and this will make it harder to understand the code.

clarissatjx commented:
In general, really clean code and I love the personalisation, especially the greeting message. The use of enums also made the main class easy to read and neat! The only thing is that in Parser.java, under the parseCommand method, the if blocks lack the curly brackets, but other than that, the naming conventions and casing are all perfect :)

verakohh committed:
Add checkStyle to detect coding style violations

verakohh committed:
Merge tag 'A-CheckStyle'

verakohh committed:
Integrate GUI chatbot
TaskList.java: Add find function and change it to accept all cases
Events.java and Deadline.java: Fix storing of tasks into file

verakohh committed:
Merge branch 'branch-Level-10'

verakohh committed:
Add varargs for delete tasks and todo tasks

verakohh committed:
Merge tag 'A-Varargs'

verakohh committed:
Change reply when task is added

yimjianbing commented on src/main/java/cleo/task/Deadline.java:
```diff
@@ -0,0 +1,63 @@
+     * @return The end time of the event in yyyy-MM-dd HH:mm format.
+     */
+    public String storeGetBy() {
+        return this.by.format(formatter);
```
Maybe you can consider removing the "this.", since java would know which "by" you're referring to, given that we are supposed to cut down as many "this." syntax as possible as part of the coding standard

yimjianbing commented on src/main/java/cleo/task/Events.java:
```diff
@@ -0,0 +1,82 @@
+     * @return The end time of the event in yyyy-MM-dd HH:mmformat.
+     */
+    public String storeGetTo() {
+        return this.to.format(formatter);
```
If possible, all the "this" can be omitted ? Since there wouldn't be a difference if they were used or not

yimjianbing commented on src/main/java/cleo/task/Events.java:
```diff
@@ -0,0 +1,82 @@
+            this.from = LocalDateTime.parse(from, formatter);
+            this.to = LocalDateTime.parse(to, formatter);
+            if (this.from.isBefore(LocalDateTime.now()) || this.to.isBefore(LocalDateTime.now())) {
+                throw new IllegalArgumentException("The date(s) cannot be in the past!");
```
Maybe you can create your own exception that extends from CleoException and throw that instead?

yimjianbing commented on src/main/java/cleo/task/Task.java:
```diff
@@ -0,0 +1,70 @@
+     *
+     * @return the value of the `isDone` flag.
+     */
+    public Boolean isDoneString() {
```
I think there may be a typo here. You are returning the wrapper class, Boolean, instead of the base boolean data type

verakohh replied:
Thank you for spotting this! I've changed it accordingly :)

yimjianbing commented on src/main/java/cleo/Storage.java:
```diff
@@ -0,0 +1,131 @@
+    /**
+     * Ensures that the data folder exists, creating it if it doesn't.
+     */
+    private void ensureDataFolderExists() {
+        File dataFolder = new File("./data");
+        if (!dataFolder.exists()) {
+            dataFolder.mkdirs(); // Create the folder and any necessary parent directories
+            System.out.println("Data folder not found. Creating './data' directory now!");
+        }
+    }
```
I really like how you abstracted out this part of the logic to create a new file if ./data doesn't exist. good job!

yimjianbing reviewed (commented):
Really good code! The java docs were done up amazingly and are really really detailed! All the naming conventions were followed, and you did a great job abstracting out the different logic, so it was really easy for me t oread through the code and understand what was going on, as well as navigate around! Amazing work!

verakohh commented:
@davidgohzk  Thank you! I've moved the `UNMARK` and `MARK` to be handled in TasksList class already :)

verakohh commented:
@clarissatjx Thank you so much!! I've added the curly brackets in Parser.java! :)))

verakohh committed:
Change GUI for greetings and add command list

verakohh committed:
Insert Assertions

Tasks have to be checked if they are converted correctly from user's
input.

Let's add assertions for Task.java to check if description and isDone is
not empty and not done respectively.
Let's check that each of the tasks are created successfully and return the
string correctly to Cleo.java.

Errors will be thrown if code fails assertions.

verakohh committed:
Refactor Cleo class for better code quality and adherence to SLAP

- Extracted methods for task handling (mark, unmark, delete)
  to improve modularity and readability.
- Updated Javadoc comments for all methods.
- Reorganised code to adhere to the Single Level of Abstraction
  Principle (SLAP), ensuring each method handles a single
  responsibility.
- Streamlined task deletion logic to handle multiple deletions
  more efficiently.

This refactor improves code clarity and maintainability,
making it easier to extend or debug the Cleo chatbot's functionality.

verakohh committed:
Merge pull request #1 from verakohh/branch-A-Assertions

Insert Assertions

verakohh committed:
Merge branch 'master' of github.com:verakohh/ip into branch-A-CodeQuality

# Conflicts:
#	data/cleo.txt
#	src/main/java/cleo/Cleo.java

verakohh committed:
Set up CI integration

verakohh committed:
Merge tag 'A-CI'

verakohh committed:
Add task priority sorting functionality and update task saving mechanism

- Let's implement `TaskPriorityComparator` to compare tasks based on
  their priority levels (P0 to P4).
- Let's add `sortByPriority` method in `TaskList` class to sort tasks by
  priority.
- Let's modify task saving mechanism in `Storage` class to include
priority levels when saving tasks to the file.
- Let's update `Cleo` chatbot logic to allow users to update task
priorities and sort tasks by priority.

With priority feature added, tasks being sorted before being listed
will be easier for users to see which tasks they should prioritise
first.

verakohh committed:
Merge branch 'C-Priority'

verakohh committed:
Change commands list and add Javadocs

verakohh committed:
Improve the GUI

GUI will show "Invalid command!" if command is not recognise or not
typed properly.

Let's add command suggestions using the Levenshtein Distance algorithm
to help users when they mistype commands. If the user enters an
invalid command, the system now suggests the closest valid command
(e.g., mistyping "deltee" will suggest "delete").

Overall, these changes improve the user interface and experience,
making it more user-friendly and error-tolerant.

verakohh committed:
Merge tag 'A-BetterGui'

verakohh committed:
Update README

verakohh committed:
Update README with hyperlinks

verakohh committed:
Update website to have light and dark mode

verakohh committed:
Update README

verakohh committed:
Update light and dark mode

verakohh committed:
Update README to have working hyperlinks

verakohh committed:
Add acknowledgment section for ChatGPT contributions in README

- Let's add a new "Acknowledgments" section to the README file to
properly credit ChatGPT by OpenAI for assisting in writing Javadocs
and designing JavaFX code.

This change ensures transparency and gives credit for external
assistance received during the development process.

verakohh committed:
Add unit tests for Parser class methods and improve error handling

Let's add unit tests for the Parser class to improve test coverage for
command parsing and error handling.

The tests ensure:
- Commands are parsed correctly.
- Invalid task numbers and priority levels trigger appropriate
  exceptions.

Improves reliability and maintains code quality by catching potential
issues early.

verakohh committed:
Refactor Storage class for improved readability and maintainability

- Let's introduce constants for task types and task statuses to replace
  magic strings.
- Let's add logging via the `Logger` class instead of using
  `System.out.println()` for error and folder creation messages.
- Let's modularise the code by extracting file creation logic into a
  separate `createNewFile()` method.
- Let's refine task-saving logic by creating a `formatTaskForStorage()`
  method to improve readability and adhere to SLAP.
- Let's implement try-with-resources in the `saveTasks()` method for
  better resource management.

This change enhances readability, makes the code easier to maintain,
and improves overall resource handling.

verakohh committed:
Refactor Cleo.java to improve readability and consistency

The Cleo.java file contains large methods that handle multiple tasks and
contain repeated logic for saving tasks after each operation.

This needs to change because the existing structure makes it harder to
read, maintain, and handle errors consistently across commands.

- Let's refactor the command handling into helper methods for each
  operation, improving readability by separating concerns.
- Let's standardise the saving of tasks, ensuring consistency after each
  command execution.
- Let's remove assertions for null checks and handle through code logic.

This approach ensures that the code is more readable, maintainable,
and easier to test in the future.

verakohh committed:
Fix bug for todo task

verakohh committed:
Fix bug for todo task and checkstyle for Cleo.java

verakohh committed:
Fix the bye command

verakohh committed:
Update code to pass smoketest

verakohh committed:
Update README.md
